### PR TITLE
dtInstance doesn't return Promise

### PR DESCRIPTION
**Pull request by shanmukhateja**
- Also updated docs and demo project

Getting this off the list before upgrade to v12

### CODE REVIEW

**Review by l-lin — COMMENTED** [review]
__issue__: I added the promise feature because there were cases where the dtOptions comes from the server. See https://l-lin.github.io/angular-datatables/#/advanced/load-dt-options-with-promise

File: src/angular-datatables.directive.ts [hunk 1]
```diff
@@ -62,62 +63,60 @@ export class DataTableDirective implements OnDestroy, OnInit {
...
-                  const instance = self.vcr.createEmbeddedView(ref, _context);
-                  self.renderer.appendChild(cellFromIndex, instance.rootNodes[0]);
-                });
-              }
+            // Filter columns using `ngTemplateRef`
+            const colsWithTemplate = columns.filter(x => x.ngTemplateRef && !x.ngPipeInstance);
+            colsWithTemplate.forEach(el => {
+              const { ref, context } = el.ngTemplateRef;
+              // get <td> element which holds data using index
+              const index = columns.findIndex(e => e.data == el.data);
```
Comment (l-lin) [inline src/angular-datatables.directive.ts:98]: __chore__: This variable `index` is shadowing the previous variable `index`. We need to rename it. Moreover, the equality must be `===` instead.

```suggestion
              const columnIndex = columns.findIndex(e => e.data === el.data);
```

File: src/angular-datatables.directive.ts [hunk 2]
```diff
@@ -62,62 +63,60 @@ export class DataTableDirective implements OnDestroy, OnInit {
...
+        rowCallback: (row, data, index) => {
+          if (this.dtOptions.columns) {
+            const columns = this.dtOptions.columns;
+            // Filter columns with pipe declared
+            const colsWithPipe = columns.filter(x => x.ngPipeInstance && !x.ngTemplateRef);
+            // Iterate
+            colsWithPipe.forEach(el => {
+              const pipe = el.ngPipeInstance;
+              // find index of column using `data` attr
+              const i = columns.findIndex(e => e.data == el.data);
```
Comment (l-lin) [inline src/angular-datatables.directive.ts:83]: __chore__: Same here with the equality.

```suggestion
              const i = columns.findIndex(e => e.data === el.data);
```

**Review by l-lin — CHANGES_REQUESTED** [review]
__issue__: The basic example "angular-way" doesn't seem to work with this change.

__suggestion__: This is a breaking change. We should not forget to add this information when releasing the newest version. Maybe add this change only in version 12, as it's a major version?

**Comment by shanmukhateja** [comment]
@l-lin thanks for the suggested changes. I've added them as requested.

And yes. I do not plan on backporting this PR to older versions.  It will be a v12+ thing only. 

We talked about this over email a few months ago, remember? I also remember attaching a git patch as POC for this.

I'm looking into angular way bug, will update when fixed.

File: src/angular-datatables.directive.ts [hunk 3]
```diff
@@ -62,62 +63,60 @@ export class DataTableDirective implements OnDestroy, OnInit {
...
-                  self.renderer.appendChild(cellFromIndex, instance.rootNodes[0]);
-                });
-              }
+            // Filter columns using `ngTemplateRef`
+            const colsWithTemplate = columns.filter(x => x.ngTemplateRef && !x.ngPipeInstance);
+            colsWithTemplate.forEach(el => {
+              const { ref, context } = el.ngTemplateRef;
+              // get <td> element which holds data using index
+              const columnIndex = columns.findIndex(e => e.data === el.data);
+              const cellFromIndex = row.childNodes.item(index);
```
Comment (l-lin) [inline src/angular-datatables.directive.ts:99]:
```suggestion
              const cellFromIndex = row.childNodes.item(columnIndex);
```

**Review by l-lin — CHANGES_REQUESTED** [review]
Sorry, my memory is quite bad.

**Comment by shanmukhateja** [comment]
Hey, I've realised this feature isn't going to work with how we setup datatables.net library.

I'm closing this for now with hope to revive it in the future, maybe Angular v13?

**Pull request closed without merging**